packages/cli/src/databases/repositories/execution.repository.ts

```diff
@@ -389,21 +399,86 @@ export class ExecutionRepository extends Repository<ExecutionEntity> {
 
 		if (!executions.length) {
 			if (deleteConditions.ids) {
-				LoggerProxy.error('Failed to delete an execution due to insufficient permissions', {
+				Logger.error('Failed to delete an execution due to insufficient permissions', {
 					executionIds: deleteConditions.ids,
 				});
 			}
 			return;
 		}
 
 		const executionIds = executions.map(({ id }) => id);
-		const binaryDataManager = BinaryDataManager.getInstance();
-		await binaryDataManager.deleteBinaryDataByExecutionIds(executionIds);
-
 		do {
 			// Delete in batches to avoid "SQLITE_ERROR: Expression tree is too large (maximum depth 1000)" error
-			const batch = executionIds.splice(0, 500);
-			await this.delete(batch);
+			const batch = executionIds.splice(0, PRUNING_BATCH_SIZE);
+			await this.softDeleteExecutions(batch);
 		} while (executionIds.length > 0);
 	}
+
+	private async pruneOlderExecutions() {
+		Logger.verbose('Pruning execution data from database');
+
+		const maxAge = config.getEnv('executions.pruneDataMaxAge'); // in h
+		const maxCount = config.getEnv('executions.pruneDataMaxCount');
+
+		// Find ids of all executions that were stopped longer that pruneDataMaxAge ago
+		const date = new Date();
+		date.setHours(date.getHours() - maxAge);
+
+		const toPrune: Array<FindOptionsWhere<IExecutionFlattedDb>> = [
+			// date reformatting needed - see https://github.com/typeorm/typeorm/issues/2286
+			{ stoppedAt: LessThanOrEqual(DateUtils.mixedDateToUtcDatetimeString(date)) },
+		];
+
+		if (maxCount > 0) {
+			const executions = await this.find({
+				select: ['id'],
+				skip: maxCount,
+				take: 1,
+				order: { id: 'DESC' },
+			});
+
+			if (executions[0]) {
+				toPrune.push({ id: LessThanOrEqual(executions[0].id) });
+			}
+		}
+
+		const executionIds = (
+			await this.find({
+				select: ['id'],
+				where: toPrune,
+				take: PRUNING_BATCH_SIZE,
+			})
+		).map(({ id }) => id);
+		await this.softDeleteExecutions(executionIds);
+
+		if (executionIds.length === PRUNING_BATCH_SIZE) {
+			setTimeout(async () => this.pruneOlderExecutions(), 1000);
+		}
+	}
+
+	private async deleteSoftDeletedExecutions() {
+		// Find ids of all executions that were deleted over an hour ago
+		const date = new Date();
+		date.setHours(date.getHours() - 1);
+
+		const executionIds = (
+			await this.find({
+				select: ['id'],
+				where: {
+					deletedAt: LessThanOrEqual(DateUtils.mixedDateToUtcDatetimeString(date)),
+				},
+				take: PRUNING_BATCH_SIZE,
+			})
+		).map(({ id }) => id);
```

Comment:
Can we delete and get the deleted IDs in the deletion result? It looks like only MySQL doesn't support `RETURNING` though...